packages/dds/tree/src/test/sequence-change-family/sequenceEditBuilder.spec.ts

```diff
@@ -0,0 +1,681 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { jsonString } from "../../domains";
+import { AnchorSet, Delta, FieldKey, UpPath } from "../../tree";
+import { SequenceEditBuilder, TextCursor } from "../../feature-libraries";
+import { brand, brandOpaque } from "../../util";
+
+const rootKey = brand<FieldKey>("root");
+const detachedKey = brand<FieldKey>("detached");
+const fooKey = brand<FieldKey>("foo");
+const barKey = brand<FieldKey>("bar");
+
+const root: UpPath = {
+    parent: undefined,
+    parentField: rootKey,
+    parentIndex: 0,
+};
+
+const detached: UpPath = {
+    parent: undefined,
+    parentField: detachedKey,
+    parentIndex: 0,
+};
+
+const root_foo2: UpPath = {
+    parent: root,
+    parentField: fooKey,
+    parentIndex: 2,
+};
+
+const root_bar2: UpPath = {
+    parent: root,
+    parentField: barKey,
+    parentIndex: 2,
+};
+
+const root_foo17: UpPath = {
+    parent: root,
+    parentField: fooKey,
+    parentIndex: 17,
+};
+
+const root_foo2_foo5: UpPath = {
+    parent: root_foo2,
+    parentField: fooKey,
+    parentIndex: 5,
+};
+
+const root_foo17_foo5: UpPath = {
+    parent: root_foo17,
+    parentField: fooKey,
+    parentIndex: 5,
+};
+
+const root_bar2_bar5: UpPath = {
+    parent: root_bar2,
+    parentField: barKey,
+    parentIndex: 5,
+};
+
+const root_foo2_foo5_foo7: UpPath = {
+    parent: root_foo2_foo5,
+    parentField: fooKey,
+    parentIndex: 7,
+};
+
+const root_bar2_bar5_bar7: UpPath = {
+    parent: root_bar2_bar5,
+    parentField: barKey,
+    parentIndex: 7,
+};
+
+const nodeX = { type: jsonString.name, value: "X" };
+const content = [nodeX];
+const moveId = brandOpaque<Delta.MoveId>(0);
+const moveId2 = brandOpaque<Delta.MoveId>(1);
+
+function test(editor: (builder: SequenceEditBuilder) => void, expected: Delta.Root): void {
+    let wasCalled = false;
+    const receiver = (actual: Delta.Root) => {
+        wasCalled = true;
+        assert.deepStrictEqual(actual, expected);
+    };
+    const anchors = new AnchorSet();
+    const builder = new SequenceEditBuilder(receiver, anchors);
+    editor(builder);
+    assert.strictEqual(wasCalled, true);
```

Comment:
I prefer this style (which I use when testing dependents):

```
    const deltas = [];
    const receiver = (delta: Delta.Root) => deltas.push(delta);
    const anchors = new AnchorSet();
    const builder = new SequenceEditBuilder(receiver, anchors);
    editor(builder);
    assert.strictDeepEqual([deltas], [expected]);
```

Reply:
If we go with this more compact, maybe we can have some or all of the tests not use this helper, and thus better replicate actual usage of SequenceEditBuilder? It would also make it easy to modify tests to intact with anchors, apply the editor in multiple steps etc.

Anyway, the high level point here is I'd like to try and make the tests not need to wrap all usages of the thing they are testing in indirection/helpers if practical. Makes the tests easier to read, serve as better examples, and easier to write tests that are slightly different. Also makes the tests test the actual API ergonomics.

Reply:
I do like the style you propose better too.

For the rest I'm less convinced (as far as this file is concerned):

Readability:
The contract for the helper is pretty straightforward, and having that code reused as a function instead of repeated everywhere means the reader is relieved of the burden of having to make sure there isn't some subtle difference in the setup code of each individual test.

Testing the API:
There seem to be two key parts to this API: (1) the methods for effecting edits and (2) the delta callback.
(1) is being tested just like it would by production code.
Is there something about the way the test helper is using (2) that is meaningfully different from what it would be if the setup code was replicated in each test?

Reply:
Its a minor subjective thing, but generally I like to be able to see the usage of the thing a test is testing when I read the body of the test, and least for the first minimal test. I don't know these APIs, and I have a hard time getting a sense for what they do from just a quick look at the tests. I can't have the data and the API use on screen at once. I can't have the editor and the SequenceEditBuilder setup on screen at once. Details like does the editor use anchors require browsing around multiple places. You can't tell any of the types in the editor call back in the tests (ex: that the builder is the SequenedEditBuilder), or even that its an "editor" when looking at the code. To read it at all I have to go review the signature of the test helper.

Also with this setup, there is no way to know when in the editor's execution the delta is sent. You can't check its sent at the right time, and I as a reader to these tests has no way to know when its supposed to actually be sent. With this setup you can't just write one more test that has an extra assert for when its sent for example. What if the editor makes multiple calls to the builder? That seems to have no tests, and with this setup its hard to add tests to cover that. Do you get multiple delta call backs? Do they get aggregated somehow? If each test is just going to make one call, but the API supports multiple, maybe we should have a table driven set of `tests call test one of each edit` then some unit tests for multiple calls.

If you really want to have a bunch of tests with identical structure and different data, and there is enough code that it shouldn't be duplicated and there is no need/desire for different tests which work slightly differently, I'd prefer a table-driven testing setup so the actual testing logic is in the test body, and we have a table to test cases as data.

I also generally see an API where you only ever want to test it with a helper like this as code smell: either we are not testing the API in all the ways it can be used, of the test helper is probably closer to the API we should actually expose.

This case isn't too bad, but I know we have built test libraries that ended up way to indirect over time in the past, and I'd like to encourage us to avoid heading in that direction when possible. It might be fine in this case, I just wanted to suggest that this pattern has costs (mainly harder to read/understand and harder to test variations of usage patterns) and suggest we consider (but not necessarily take) alternative approaches.

Reply:
Thanks for elaborating on that.

I can see how the helper is hiding some of the control flow here. We could have something like this:
```typescript
function setup(anchors?: AnchorSet): { builder: SequenceEditBuilder; deltas: Delta.Root[]; } {
    const deltas: Delta.Root[] = [];
    const receiver = (delta: Delta.Root) => deltas.push(delta);
    const builder = new SequenceEditBuilder(receiver, anchors ?? new AnchorSet());
    return { builder, deltas };
}
```

Which would then make the first test this:
```typescript

it("Can set the root node value", () => {
    const { builder, deltas } = setup();
    builder.setValue(root, 42);
    const expected: Delta.Root[] = [new Map([[
        rootKey,
        [{
            type: Delta.MarkType.Modify,
            setValue: 42,
        }],
    ]])];
    assert.deepStrictEqual(deltas, expected);
});
```

I'm guessing it may still be too obfuscated for your taste, but I'm interested to hear whether you think it's a step in the right direction.

About "I don't know these APIs, and I have a hard time getting a sense for what they do from just a quick look at the tests.":
Did the above address the part that was hard to infer from the test? If not, what part of the contract is unclear from the proposed change?

About "You can't tell any of the types in the editor call back in the tests":
Are you no looking at the PRs in VSCode? My VSCode includes that type as an inlay hint.

About "Details like does the editor use anchors require browsing around multiple places":
The proposed code above tries to address that, let me know if you're happy with that aspect.

About what to expect when multiple calls are made to the builder:
Each call will lead to a delta being sent. The reason I had no tests for that is because we don't yet have support for squashing the multiple changesets being produced into a single changeset. Looking into this more it turns out that the squashing logic does not get invoked even when multiple changes are made, so that means I can and should add tests for that.

About "I'd prefer a table-driven testing setup so the actual testing logic is in the test body, and we have a table to test cases as data.":
Yeah I like that idea, I think I'm going to try that here.

About "I also generally see an API where you only ever want to test it with a helper like this as code smell":
Yes! I really like that point and I'm thankful you're here to remind me of it. This has me wondering, would the DevX of `ProgressiveEditBuilder` be improved if we had it return a `Delta.Root` on each edit instead of having this `deltaReceiver` pattern? It would make the tests simpler, but I want to be cautious about the fact that the actual production usage of the class being different here.

Reply:
I think the table-driven testing setup will be awkward here:
- Anchors can be added or removed any time and checked any time. Having a format for the test input and expected output that's able to add and check specific anchors out of a set will make the format obscure.
- The table format would have to represent calls to be made to the builder. Using lambdas is the simplest and most flexible but it doesn't communicate the "one builder call => one delta" structure. Using records instead would help with that but would make the test code less like the actual API usage.

Given all that I'm going to go for the pattern shown in my previous comment.

Reply:
In the end I felt the setup function didn't add enough value, so I'm just using the API directly.